### ISSUE TITLE
depends-on / linked containers not working

### ISSUE THREAD

**Issue by wehrstedt** [containrrr/watchtower#1266]
**Describe the bug**
I tried to use the linked container feature to make sure my containers are stopped in the correct order before watchtower updates the services. I tried multiple ways of defining the `com.centurylinklabs.watchtower.depends-on` label, but the containers are always stopped in the same (wrong) order.

**To Reproduce**
I created the following docker-compose file:
```
version: "3"
services:

  service1:
    image: image-1:latest
    labels:
      com.centurylinklabs.watchtower.enable: "true"

  service2:
    image: image-2:latest
    depends_on:
      - service3
    labels:
      com.centurylinklabs.watchtower.depends-on: "service3"
      com.centurylinklabs.watchtower.enable: "true"

  service3:
    image: image-3:latest
    depends_on:
      - service1
    labels:
      com.centurylinklabs.watchtower.depends-on: "service1"
      com.centurylinklabs.watchtower.enable: "true"
```

I found multiple discussions and issues about that feature with different ways of defining the label. These are the versions I tried:
* ```yml
  labels:
    com.centurylinklabs.watchtower.depends-on: "service3"
  ```
* ```yml
  labels:
    com.centurylinklabs.watchtower.depends-on: "/service3" # found that in an other issue
  ```
* ```yml
  labels:
    com.centurylinklabs.watchtower.depends-on: "watchtower-test_service3-1" # tried the full container name
  ```
* ```yml
  labels:
    - com.centurylinklabs.watchtower.depends-on: "service3" # using an array
  ```

and some variants of the versions above. The containers are always updated / recreated in the following order:
* stopping service1
* stopping service2
* stopping service3
* creating service3
* creating service2
* creating service1

**Expected behavior**
The services should be updated / recreated in the following order:
* stopping service2
* stopping service3
* stopping service1
* creating service1
* creating service3
* creating service2

**Environment**
<!-- 
We want to know:
 - Platform Ubuntu 18.04.6 LTS
 - Architecture
 - Docker version Docker version 20.10.12, build e91ed57
 - ```docker image inspect containrrr/watchtower
[
    {
        "Id": "sha256:333de6ea525af9137e1f14a5c1bfaa2e730adca97ab97f74d738dfa99967f14f",
        "RepoTags": [
            "containrrr/watchtower:latest"
        ],
        "RepoDigests": [
            "containrrr/watchtower@sha256:bbf9794a691b59ed2ed3089fec53844f14ada249ee5e372ff0e595b73f4e9ab3"
        ],
        "Parent": "",
        "Comment": "buildkit.dockerfile.v0",
        "Created": "2022-01-22T18:09:08.462846285Z"
```
-->

**Comment by piksel** [comment]
```yaml
labels:
  com.centurylinklabs.watchtower.depends-on: "watchtower-test_service3-1" # tried the full container name
```

This is probably the only one that would work. 
The container name is what is reported by `docker ps`. The project prefix is compose-specific and not something that watchtower knows about.
The order of the stopping and starting looks correct otherwise.

**Comment by wehrstedt** [comment]
Okay, I retried this with an minimal example and finally get it to work. You are right, you have to use the full container name. But you also have to add a leading slash. Here are my results with and without leading slash:

watchtower
```yml
version: "3"
services:

  watchtower:
    image: containrrr/watchtower
    container_name: "watchtower2"
    restart: always
    command: --run-once
    environment:
      - WATCHTOWER_DEBUG=true
      - WATCHTOWER_CLEANUP=false
      - WATCHTOWER_LABEL_ENABLE=true
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - /etc/localtime:/etc/localtime:ro
      - /home/ubuntu/.docker/config.json:/config.json:ro
```

I created a small bash-script to re-tag the images after each watchtower run:
```
#!/bin/bash

# Pull a version and re-tag them to latest
docker pull mariadb:10.6
docker pull nginx:stable-alpine
docker pull traefik:v2.7

docker tag mariadb:10.6 mariadb:latest
docker tag nginx:stable-alpine nginx:latest
docker tag traefik:v2.7 traefik:latest

# re-create the stack
docker-compose down
docker-compose up -d
```

### Without leading slash
compose-file
```yml
version: "3"
services:

  service1:
    image: mariadb:latest
    container_name: "service1"
    environment:
      MARIADB_RANDOM_ROOT_PASSWORD: "true"
    labels:
      com.centurylinklabs.watchtower.enable: "true"

  service2:
    image: nginx:latest
    container_name: "service2"
    labels:
      com.centurylinklabs.watchtower.enable: "true"
      com.centurylinklabs.watchtower.depends-on: "service3"

  service3:
    image: traefik:latest
    container_name: "service3"
    labels:
      com.centurylinklabs.watchtower.enable: "true"
      com.centurylinklabs.watchtower.depends-on: "service1"
```

result
```
watchtower2   | time="2022-04-12T10:23:44+02:00" level=info msg="Stopping /service3 (3ead0d81a056) with SIGTERM"
watchtower2   | time="2022-04-12T10:23:45+02:00" level=debug msg="Removing container 3ead0d81a056"
watchtower2   | time="2022-04-12T10:23:45+02:00" level=info msg="Stopping /service1 (093c70be7021) with SIGTERM"
watchtower2   | time="2022-04-12T10:23:46+02:00" level=debug msg="Removing container 093c70be7021"
watchtower2   | time="2022-04-12T10:23:46+02:00" level=info msg="Stopping /service2 (f7bd88cd9446) with SIGTERM"
watchtower2   | time="2022-04-12T10:23:47+02:00" level=debug msg="Removing container f7bd88cd9446"
watchtower2   | time="2022-04-12T10:23:47+02:00" level=info msg="Creating /service2"
watchtower2   | time="2022-04-12T10:23:48+02:00" level=debug msg="Starting container /service2 (212fff269e66)"
watchtower2   | time="2022-04-12T10:23:48+02:00" level=info msg="Creating /service1"
watchtower2   | time="2022-04-12T10:23:49+02:00" level=debug msg="Starting container /service1 (ee9c69533a01)"
watchtower2   | time="2022-04-12T10:23:50+02:00" level=info msg="Creating /service3"
watchtower2   | time="2022-04-12T10:23:50+02:00" level=debug msg="Starting container /service3 (2944aba20d2d)"
```

### With leading trash
compose-file
```yml
version: "3"
services:

  service1:
    image: mariadb:latest
    container_name: "service1"
    environment:
      MARIADB_RANDOM_ROOT_PASSWORD: "true"
    labels:
      com.centurylinklabs.watchtower.enable: "true"

  service2:
    image: nginx:latest
    container_name: "service2"
    labels:
      com.centurylinklabs.watchtower.enable: "true"
      com.centurylinklabs.watchtower.depends-on: "/service3"

  service3:
    image: traefik:latest
    container_name: "service3"
    labels:
      com.centurylinklabs.watchtower.enable: "true"
      com.centurylinklabs.watchtower.depends-on: "/service1"
```

results
```
watchtower2   | time="2022-04-12T10:31:14+02:00" level=info msg="Stopping /service2 (633de7cb4250) with SIGTERM"
watchtower2   | time="2022-04-12T10:31:15+02:00" level=debug msg="Removing container 633de7cb4250"
watchtower2   | time="2022-04-12T10:31:15+02:00" level=info msg="Stopping /service3 (51870867dfd4) with SIGTERM"
watchtower2   | time="2022-04-12T10:31:16+02:00" level=debug msg="Removing container 51870867dfd4"
watchtower2   | time="2022-04-12T10:31:16+02:00" level=info msg="Stopping /service1 (d55f6fa1f941) with SIGTERM"
watchtower2   | time="2022-04-12T10:31:17+02:00" level=debug msg="Removing container d55f6fa1f941"
watchtower2   | time="2022-04-12T10:31:17+02:00" level=info msg="Creating /service1"
watchtower2   | time="2022-04-12T10:31:17+02:00" level=debug msg="Starting container /service1 (2783408a8fb5)"
watchtower2   | time="2022-04-12T10:31:18+02:00" level=info msg="Creating /service3"
watchtower2   | time="2022-04-12T10:31:18+02:00" level=debug msg="Starting container /service3 (238a027eb1f0)"
watchtower2   | time="2022-04-12T10:31:19+02:00" level=info msg="Creating /service2"
watchtower2   | time="2022-04-12T10:31:19+02:00" level=debug msg="Starting container /service2 (911129d3236c)"
```

In summary, using the full container name with leading slash solves my problem. But I noticed some strange side effects testing this which I could solve starting watchtower with `docker-compose up --force-recreate`. Without `--force-recreate`, watchtower sometimes stopped immediately with the message `No new image for X found`, even I re-tagged the images with bash-script above and re-deployed the stack with the old image versions. After adding `--force-recreate`, watchtower detected the new image versions as expected.

Maybe there is some kind of cache which gets not invalidated in some situations, what would explain why it was not possible to  reproduce / verify the error (like mentioned in https://github.com/containrrr/watchtower/pull/1103)?

**Comment by wehrstedt** [comment]
> The container name is what is reported by `docker ps`. The project prefix is compose-specific and not something that watchtower knows about.

The compose project prefix is available as a container label which can be retrieved by `docker inspect`:
```json
            "Labels": {
                "com.centurylinklabs.watchtower.enable": "true",
                "com.docker.compose.config-hash": "0f530eb3cb372b60d82621502ff5d928ef5d0a7fb3cd00b3034ceb6d62b39c06",
                "com.docker.compose.container-number": "1",
                "com.docker.compose.oneoff": "False",
                "com.docker.compose.project": "watchtower-test",
                "com.docker.compose.service": "service3",
                "com.docker.compose.version": "1.17.1"
```

Would be possible to generate the full container name with these labels.

**Comment by piksel** [comment]
The leading slashes should not be needed as there is an explicit check for them:
https://github.com/containrrr/watchtower/blob/d12ce7ce79009a62a36b0abd36ac0bfd19a71fa3/internal/actions/update.go#L246-L249

> Maybe there is some kind of cache which gets not invalidated in some situations  

watchtower itself is stateless, but it might be due to some state that docker-compose keeps?

> Would be possible to generate the full container name with these labels.

Sure, but no one has made such an implementation yet. For now we are totally compose-unaware.

**Comment by piksel** [comment]
Also, there should be debug messages that tell you how the containers are linked:
https://github.com/containrrr/watchtower/blob/d12ce7ce79009a62a36b0abd36ac0bfd19a71fa3/internal/actions/update.go#L231-L234

**Comment by ghost** [comment]
> The leading slashes should not be needed as there is an explicit check for them:

I have the same problem that I need to add slashes in labels to get the right stop/create order.

Example compose file:
```
services:
  miniflux:
    image: miniflux/miniflux:latest
    container_name: miniflux
    restart: unless-stopped
    ports:
      - "8080:8080"
    depends_on:
      - miniflux-db
    environment:
      DATABASE_URL: postgres://user:password@miniflux-db/miniflux?sslmode=disable
      RUN_MIGRATIONS: 1
      CREATE_ADMIN: 1
      ADMIN_USERNAME: user
      ADMIN_PASSWORD: password
    labels:
      com.centurylinklabs.watchtower.depends-on: /miniflux-db
  miniflux-db:
    image: postgres:15
    container_name: miniflux-db
    restart: unless-stopped
    environment:
      POSTGRES_DB: miniflux
      POSTGRES_USER: user
      POSTGRES_PASSWORD: password
    volumes:
      - data:/var/lib/postgresql/data

volumes:
  data:
```

Without slashes (stop database first, then frontend):
```
time="2022-11-12T04:00:20+08:00" level=info msg="Found new postgres:15 image (6e5b7c4abf29)"
time="2022-11-12T04:00:33+08:00" level=info msg="Stopping /miniflux-db (3ee8297661a8) with SIGTERM"
time="2022-11-12T04:00:34+08:00" level=info msg="Stopping /miniflux (29167709e942) with SIGTERM"
time="2022-11-12T04:00:46+08:00" level=info msg="Creating /miniflux"
time="2022-11-12T04:00:47+08:00" level=info msg="Creating /miniflux-db"
```

With slashes (stop frontend first, then database):
```
time="2022-11-16T04:00:30+08:00" level=info msg="Found new postgres:15 image (68f5d950dcd3)"
time="2022-11-16T04:01:15+08:00" level=info msg="Stopping /miniflux (ee91211f9ae1) with SIGTERM"
time="2022-11-16T04:01:16+08:00" level=info msg="Stopping /miniflux-db (4e86b729c335) with SIGTERM"
time="2022-11-16T04:01:17+08:00" level=info msg="Creating /miniflux-db"
time="2022-11-16T04:01:17+08:00" level=info msg="Creating /miniflux"
```

**Comment by ribbons** [comment]
> The leading slashes should not be needed as there is an explicit check for them:

I've never used go in anger (so please disregard this if I'm barking up the wrong tree), but the package sorting code doesn't appear to prepend a slash when it is evaluating the container order:

https://github.com/containrrr/watchtower/blob/29f5c4b254a686d9198277a78f5066666240f0e0/pkg/sorter/sort.go#L71-L73

https://github.com/containrrr/watchtower/blob/29f5c4b254a686d9198277a78f5066666240f0e0/pkg/sorter/sort.go#L87-L90

That would seem to explain the behaviour that @whaler-ragweed, @wehrstedt and I have observed that the containers are not restarted in the correct order if the depends-on label doesn't include a leading slash.

**Comment by piksel** [comment]
@ribbons Yeah, that might be the case. I tried to follow the lookups to see if it shouldn't be needed for some reason, but I think you are right. It would mean that the containers get included for restarting, but not in the correct order (which matches what you have experienced).
